test/e2e_env/kubernetes/gateway/cross-mesh.go

```diff
@@ -0,0 +1,178 @@
+package gateway
+
+import (
+	"fmt"
+
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/kumahq/kuma/test/e2e_env/kubernetes/env"
+	. "github.com/kumahq/kuma/test/framework"
+	"github.com/kumahq/kuma/test/framework/client"
+	"github.com/kumahq/kuma/test/framework/deployments/testserver"
+)
+
+func CrossMeshGatewayOnKubernetes() {
+	// ClientNamespace is a namespace to deploy gateway client
+	// applications. Mesh sidecar injection is not enabled there.
+	const gatewayClientNamespace = "cross-mesh-kuma-client"
+	const gatewayClientNamespaceOtherMesh = "cross-mesh-kuma-client2"
+	const gatewayTestNamespace = "cross-mesh-kuma-test"
+
+	const echoServer = "cross-mesh-echo-server"
+	const edgeGateway = "cross-mesh-edge-gateway"
+
+	const gatewayMesh = "cross-mesh-gateway"
+	const gatewayOtherMesh = "cross-mesh-other"
+
+	const gatewayPort = 8080
+
+	echoServerApp := testserver.Install(
+		testserver.WithMesh(gatewayMesh),
+		testserver.WithName(echoServer),
+		testserver.WithNamespace(gatewayTestNamespace),
+		testserver.WithArgs("echo", "--instance", "kubernetes"),
+	)
+
+	// DeployCluster creates a Kuma cluster on Kubernetes using the
+	// provided options, installing an echo service as well as a
+	// gateway and a client container to send HTTP requests.
+	DeployCluster := func() {
+		setup := NewClusterSetup().
+			Install(MTLSMeshKubernetes(gatewayMesh)).
+			Install(MTLSMeshKubernetes(gatewayOtherMesh)).
+			Install(NamespaceWithSidecarInjection(gatewayTestNamespace)).
+			Install(NamespaceWithSidecarInjection(gatewayClientNamespace)).
+			Install(NamespaceWithSidecarInjection(gatewayClientNamespaceOtherMesh)).
+			Install(echoServerApp).
+			Install(DemoClientK8s(gatewayOtherMesh, gatewayClientNamespace)).
+			Install(DemoClientK8s(gatewayMesh, gatewayClientNamespaceOtherMesh))
+
+		Expect(setup.Setup(env.Cluster)).To(Succeed())
+	}
+
+	// Before each test, install the gateway and routes.
+	JustBeforeEach(func() {
+		Expect(k8s.KubectlApplyFromStringE(
+			env.Cluster.GetTesting(),
+			env.Cluster.GetKubectlOptions(gatewayTestNamespace), fmt.Sprintf(`
```

Comment:
I prefer `YamlK8s` like in other tests. You can then also chain this in ClusterSetup